src/whir/utils.rs

```diff
@@ -36,42 +36,123 @@ pub const fn workload_size<T: Sized>() -> usize {
+/// After folding reduces the Reed–Solomon codeword domain, this function selects random
+/// query positions for the proximity test. Uses Fiat-Shamir randomness to ensure unpredictable,
+/// verifier-determined challenge locations.
 ///
 /// ## Parameters
-/// - `domain_size`: The size of the original evaluation domain (e.g., 2^22).
-/// - `folding_factor`: The number of folding rounds applied (e.g., k = 1 means domain halves).
-/// - `num_queries`: The number of query *indices* we want to obtain.
-/// - `challenger`: A Fiat–Shamir transcript used to sample randomness deterministically.
+/// - `domain_size`: Original evaluation domain size before folding
+/// - `folding_factor`: Number of folding rounds (domain reduction = 2^folding_factor)
+/// - `num_queries`: Target number of query indices to sample
+/// - `prover_state`: Fiat-Shamir transcript state for deterministic randomness
 ///
 /// ## Returns
-/// A sorted and deduplicated list of random query indices in the folded domain.
-pub fn get_challenge_stir_queries<Chal: ChallengeSampler<EF>, F, EF>(
+/// Sorted, deduplicated vector of query indices in [0, folded_domain_size)
+pub fn get_challenge_stir_queries<Challenger, F, EF>(
     domain_size: usize,
     folding_factor: usize,
     num_queries: usize,
-    prover_state: &mut Chal,
+    prover_state: &mut Challenger,
 ) -> Result<Vec<usize>, FiatShamirError>
 where
+    Challenger: ChallengeSampler<EF>,
     F: Field,
     EF: ExtensionField<F>,
 {
-    // Folded domain size = domain_size / 2^folding_factor.
+    // PHASE 1: Compute folded domain parameters
+
+    // Apply WHIR folding: domain_size / 2^folding_factor
+    //
+    // Example: 1M domain → 3 folds → 1M >> 3 = 125K domain
     let folded_domain_size = domain_size >> folding_factor;
 
-    // Number of bits needed to represent an index in the folded domain.
+    // Bits required to index into the folded domain
+    //
+    // Example: 125K domain needs ceil(log_2(125K)) ≈ 17 bits per query
     let domain_size_bits = log2_ceil_usize(folded_domain_size);
 
-    // Sample one integer per query, each with domain_size_bits of entropy.
-    let queries = (0..num_queries)
-        .map(|_| prover_state.sample_bits(domain_size_bits) % folded_domain_size)
-        .sorted_unstable()
-        .dedup()
-        .collect();
+    // PHASE 2: Determine batching strategy
+
+    // The maximum number of bits that can be safely sampled in a single call from the challenger.
+    //
+    // To avoid statistical bias when sampling randomness, the number of bits requested should be
+    // less than the bit length of the challenger's underlying field modulus.
+    let max_bits_per_call = F::bits() - 1;
+
+    // Total entropy needed: num_queries × bits_per_query
+    //
+    // Example: 100 queries × 4 bits = 400 bits total
+    let total_bits_needed = num_queries * domain_size_bits;
+
+    // PHASE 3: Execute sampling
+    let queries = if total_bits_needed > 0 && domain_size_bits > 0 {
```

Comment:
`domain_size_bits == 0` implies that `total_bits_needed == 0` so you only need to check the first.
```suggestion
    let queries = if total_bits_needed > 0 {
```